### ISSUE TITLE
k3s kilo pods crashlooping

### ISSUE THREAD

**Issue by ibalajiarun** [squat/kilo#27]
> sudo kubectl apply -f https://raw.githubusercontent.com/squat/kilo/master/manifests/kilo-k3s-flannel.yaml
```
serviceaccount/kilo created
clusterrole.rbac.authorization.k8s.io/kilo created
clusterrolebinding.rbac.authorization.k8s.io/kilo created
daemonset.apps/kilo created
```
> sudo kubectl logs -f kilo-cz64w -n kube-system

```failed to create Kubernetes config: Error loading config file "/etc/kubernetes/kubeconfig": read /etc/kubernetes/kubeconfig: is a directory```

I think the problem is with `kilo-k3s-flannel.yaml:99`.

**Comment by ibalajiarun** [comment]
I fixed the config to point to the right file, but I am running into this new issue:

```
failed to create Kubernetes config: invalid configuration: [unable to read client-cert /var/lib/rancher/k3s/agent/client-kubelet.crt for user due to open /var/lib/rancher/k3s/agent/client-kubelet.crt: no such file or directory, unable to read client-key /var/lib/rancher/k3s/agent/client-kubelet.key for user due to open /var/lib/rancher/k3s/agent/client-kubelet.key: no such file or directory, unable to read certificate-authority /var/lib/rancher/k3s/agent/server-ca.crt for local due to open /var/lib/rancher/k3s/agent/server-ca.crt: no such file or directory]```

**Comment by squat** [comment]
Hi @ibalajiarun, what version of k3s are you running?
Also, what change did you make to the config file? It would be helpful to know what file you are trying to load.

**Comment by ibalajiarun** [comment]
Hi @squat, I got it working after re-doing all the steps in #11 with some modifications.
But, I ended up using `kilo-k3s.yaml` because `kilo-k3s-flannel.yaml` did not work for some reason. I will try again today.
In either files, I had to update the `kubeconfig` volume path:
```
❯ git diff manifests/kilo-k3s.yaml
diff --git a/manifests/kilo-k3s.yaml b/manifests/kilo-k3s.yaml
index 34c0572..6b8b5aa 100644
--- a/manifests/kilo-k3s.yaml
+++ b/manifests/kilo-k3s.yaml
@@ -158,4 +158,4 @@ spec:
           path: /var/lib/kilo
       - name: kubeconfig
         hostPath: 
-          path: /var/lib/rancher/k3s/agent/kubeconfig.yaml
+          path: /etc/rancher/k3s/k3s.yaml
```

I also had to copy /etc/rancher/k3s/k3s.yaml to the other nodes, because I dont think they are copies anymore as you say in #11.

**Comment by squat** [comment]
I’m glad you got it working. To help figure out what was going out, what
version of k3s are you using?

**Comment by ibalajiarun** [comment]
k3s v1.0.1

**Comment by squat** [comment]
Hi @ibalajiarun sorry i haven't tried spinning up a new k3s node since you filled the issue :/

Would you mind making a PR with the corrected paths for the kubeconfig? I'll try to launch an new k3s cluster this weekend to validate. Thanks for your time :tada: 

-Lucas

**Comment by p-hash** [comment]
Hello. I'm having the same issue. 

I've tested with k3s versions v1.0.0 and v1.17.3+k3s1. Both with both kilo-k3s.yaml and kilo-k3s-flannel.yaml. The issue persists in all 4 cases.

I cannot fully understand how should I correct the paths. 

Also this brings up a question: how to make sure kilo works with different versions of k3s? And is compatibility with older versions really an issue?

**Comment by squat** [comment]
Hi @p-hash can you share your logs just to make sure?

Regarding older versions of k3s, unfortunately the filesystem layout changed quite a bit in the last while before 1.0 and we can't expect to have manifests for every single version of every installer :/ I think a good goal would be to aim for the manifets to be compatible with the latest stable release.

**Comment by p-hash** [comment]
```
[vagrant@node1 ~]$ k3s --version
k3s version v1.17.3+k3s1 (5b17a175)
[vagrant@node1 ~]$ k3s kubectl apply -f https://raw.githubusercontent.com/squat/kilo/master/manifests/kilo-k3s.yaml
configmap/kilo created
serviceaccount/kilo created
clusterrole.rbac.authorization.k8s.io/kilo created
clusterrolebinding.rbac.authorization.k8s.io/kilo created
daemonset.apps/kilo created
[vagrant@node1 ~]$ k3s kubectl -n kube-system get pod
NAME                                      READY   STATUS             RESTARTS   AGE
local-path-provisioner-58fb86bdfd-g7zn9   1/1     Running            0          63s
kilo-xqpv6                                0/1     CrashLoopBackOff   1          41s
coredns-d798c9dd-s8p5w                    1/1     Running            0          63s
[vagrant@node1 ~]$ k3s kubectl -n kube-system logs kilo-xqpv6
failed to create Kubernetes config: Error loading config file "/etc/kubernetes/kubeconfig": read /etc/kubernetes/kubeconfig: is a directory
```

I can give you logs for other configurations (`k3s v1.0.0` and/or `kilo-k3s-flannel.yaml`) if you need them. Basically the error is the same for all of them.

Thanks for answer about older versions. Having it working with release v1.0.0 is obviously more important.

Any idea on which kubeconfig to use? There are 3:
```
[root@node1 agent]# pwd
/var/lib/rancher/k3s/agent
[root@node1 agent]# find -name '*kubeconfig*' -type f
./kubelet.kubeconfig
./kubeproxy.kubeconfig
./k3scontroller.kubeconfig
```

**Comment by p-hash** [comment]
All of those `*.kubeconfig` files need corresponding certificates for operation. I will try to fix manifest and let you know if i get any progress

**Comment by ibalajiarun** [comment]
@p-hash Make sure to also copy `/etc/rancher/k3s/k3s.yaml` from master to all the nodes. They are no longer make available automatically by k3s.

**Comment by p-hash** [comment]
Was able to run kilo manifest with `/etc/rancher/k3s/k3s.yaml` as kubeconfig. Thanks @ibalajiarun

**Comment by squat** [comment]
@p-hash would you that new manifest as a PR? At least so the config works for the master nodes? :)

**Comment by squat** [comment]
Closing this thanks to#48

**Comment by p-hash** [comment]
The solution works only for master. Kilo pods are still crashing on worker nodes. 

I think the issue should be reopen until the problem is fixed or at least addressed in documentation

**Comment by squat** [comment]
Sure @p-hash :+1: 
I think the best way forward in the short term is to add a comment in the k3s manifests that said eg "this kubeconfig must be copied onto the worker nodes". WDYT?
This issue is actually very closely related to #49. The reason we need to grab the highly privileged kubeconfig in the first place, rather than use the service account that is configured in the manifest, is that the kubeconfig contains a server address with a real, non-ServiceCIDR IP, for the API. Kilo cannot generally use a service IP to connect to the API server because Kilo itself is responsible for creating the network fabric, and the service IPs will not be routable before Kilo is it up. The ideal path forward, which we need to implement in order to close 49, is to use the credentials from the configured service account but to take the external server address from a kubeconfig.

**Comment by adamkpickering** [comment]
As I understand, the creation of the `rest.Config` object used by kilo to talk to the API currently goes as follows:

1. `kubeconfig` (the path to the kubeconfig) and `master` (the IP address of the master node) are read in from command line flags if the user specified them

2. `kubeconfig` and `master` are passed into the `clientcmd.BuildConfigFromFlags` function. Neither of these arguments are required (though I believe there is a bug in this function: if you pass only `master` in, but no `kubeconfig`, isn't there nothing to override?). If you pass in both, the `rest.Config` is created from the kubeconfig referenced by `kubeconfig`, and `rest.Config.Host` is filled from the value of `master`. In this case, kilo doesn't use the service account and cluster role that we create for it in each manifest; it uses the client certificate from the kubeconfig for authentication and authorization, giving it the ability to do anything the admin can.

3. If we don't pass `kubeconfig` and `master` to `clientcmd.BuildConfigFromFlags`, it falls back to using `rest.InClusterConfig`, which first reads the service account token from its standard location, and then reads the `KUBERNETES_SERVICE_HOST` and `KUBERNETES_SERVICE_PORT` environment variables to get the location of the api server. I'm not sure how the kubelet sets these variables, but in my installation `KUBERNETES_SERVICE_HOST` is a service IP, so it's only reachable from within the cluster.

TLDR: Assuming that in all use cases we have a service account, we should be able to override the `KUBERNETES_SERVICE_HOST` and `KUBERNETES_SERVICE_PORT` environment variables according to flags the user passes in, in the event that we need to run kilo before we have a network fabric running. If we *do* have a network fabric running, the user doesn't have to worry about where the api server is - it's taken care of by the kubelet when it sets these environment variables to point to the api server service. This avoids the need to copy kubeconfigs around, or pass sensitive information to kilo. What do you think?

I have to admit that I don't fully understand how the kubelet sets these environment variables inside a pod. I did some snooping around flannel's code, and it seems to assume that in a k8s environment these environment variables will always be set. So wouldn't it be reasonable to make the same assumption in kilo, given that it does a very similar thing? According to some other [reading](https://medium.com/containerum/4-ways-to-bootstrap-a-kubernetes-cluster-de0d5150a1e4) I did, the CNI plugin is totally optional - I would guess that without a network fabric, the kubelet, which must be configured with the api server's public IP in this case, sets `KUBERNETES_SERVICE_HOST` as the public IP of the api server. Are you familiar with the details here @squat ?

**Comment by tetricky** [comment]
I'm having this problem with k3s (v1.18.8+k3s1) on debian buster with iptables (and ufw). Single master, two worker nodes (KVM vm's networked with NAT, can reach each other on the subnet, 6443 and 51820 open on all nodes). No traefik, metallb, or flannel (I'm aiming to get a bare cluster, based on kilo networking). The kilo manifest points to "kubeconfig=/etc/kubernetes/kubeconfig" (from the kilo-k3s.yaml). Which doesn't exist on any node. On the server kubeconfig is at /etc/rancher/k3s/k3s.yaml the server pod runs, but log has:

{"caller":"cni.go:58","component":"kilo","err":"failed to read IPAM config from CNI config list file: no IP ranges specified","level":"warn","msg":"failed to get CIDR from CNI file; overwriting it","ts":"2020-09-14T18:52:31.564522566Z"}
{"caller":"cni.go:66","component":"kilo","level":"info","msg":"CIDR in CNI file is empty","ts":"2020-09-14T18:52:31.56461313Z"}
{"CIDR":"10.42.0.0/24","caller":"cni.go:71","component":"kilo","level":"info","msg":"setting CIDR in CNI file","ts":"2020-09-14T18:52:31.564666128Z"}

On the worker nodes  /etc/rancher/k3s/k3s.yaml appears to be an empty directory. and the pods running on the nodes are "CrashLoopBackOff" and log has:

failed to create Kubernetes config: Error loading config file "/etc/kubernetes/kubeconfig": read /etc/kubernetes/kubeconfig: is a directory

I am unclear from this issue what the suggestion is about copying kubeconfig (k3s.yaml) to where, and whether I need to edit to point to the server ip from the worker nodes, or whether 127.0.0.0 should work for all.

Apologies if this is an offensive monumental lack of understanding, because it is...I'm fishing around in the foothills of my own knowledge, trying to get this working, and clearly have little idea what I'm doing.

I'm working off https://jbhannah.net/articles/k3s-wireguard-kilo with some mods (addition of iptables to debian buster, etc).

My interest in kilo is that I see great application in connecting a centralised cluster, with individual edge single node clusters, using encryption, to share and route services. Thank you in advance for any guidance.

**Comment by squat** [comment]
hi @tetricky, in the future, let's tackle these issues in a new issue :))
regarding the problem you're experiencing, please take note of the comment in the k3s manifest: https://github.com/squat/kilo/blob/master/manifests/kilo-k3s.yaml#L169-L172

We need to copy the k3s.yaml file from the master to the same place on the worker nodes. In your case, that means you should:
1. delete the Kilo daemonset;
1. delete the `/etc/rancher/k3s/k3s.yaml` from the worker nodes;
1. make sure the k3s.yaml file from the master node uses a server address that is routable from the internet, e.g. an externally resolvable DNS name that resolves to a public IP address;
1. copy the file to the worker nodes; and then 
1. reapply the kilo manifests to the cluster.

**Comment by tetricky** [comment]
3. Is definitely an issue. Given that these are VM's using NAT on kvm, does that mean that I need to bridge instead (problematic), or can i port forward (6443?) from the kvm host (which does have a public ip)?

**Comment by squat** [comment]
Aha, I see. Are these all VMs on the same machine? If so, then localhost or 127.0.0.1 is likely ok. Kilo doesn't care if the API endpoint is 127.0.0.1 or a DNS name or anything else. The only important thing is that the endpoint (host+port) be accessible from every worker node before the VPN is established, this is why the endpoint should typically be a public IP, as the nodes are often in different cloud providers or networks

**Comment by tetricky** [comment]
Yes, in my specific use case the VM's are all on the same machine and the same (NAT) subnet. My intention is to have services on this kvm host, which then allows external connections over kilo to access these services from other clusters (as well as expose them through the kvm host).

**Comment by squat** [comment]
Ok, that's clear to me now :) sounds like you have a path forward. If you run into more issues, or need help, let's follow up in a new issue particular to this special use-case, or we can chat in the kubernetes slack :)

**Comment by squat** [comment]
Closing this issue as it was a documentation problem. The future discussion and development of enabling Kilo to operate without an admin Kubeconfig will be tracked in #49 :)
